### PR TITLE
Ship reconciler types

### PR DESCRIPTION
**Pull request by prometheansacrifice**
I couldn't get `flow gen-flow-files` to extract all the types. Ended up manually putting together some of the types from `ReactFiberReconciler.js`. How can I make this better?

Also, if there's no way to do this via flow or any other tool and are forced to maintain it separately I was wondering if we could test these types like `flow-typed` currently does - use [flow-typed's custom eslint rules](https://github.com/jeffmo/eslint-plugin-flow-typed/) and write a test reconciler (both passing and failing types) just like how flow-typed expects lib authors to write tests.

### CODE REVIEW

**Comment by prometheansacrifice** [comment]
cc @gaearon

**Comment by jquense** [comment]
Graphql exposes it's types without separate type files in the repo, maybe we can see how they build the flow.js files and copy that approach?

**Comment by prometheansacrifice** [comment]
@jquense Looks like they copy the all source with a `.flow` extension. I don't see any extraction of types. There are two copies of the source: one compiled down, and the other one being the original file  renamed to .flow

instanceof.js.flow
```
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/**
 * A replacement for instanceof which includes an error warning when multi-realm
 * constructors are detected.
 */
declare function instanceOf(
  value: mixed,
  constructor: mixed,
): boolean %checks(value instanceof constructor);

// eslint-disable-next-line no-redeclare
export default function instanceOf(value, constructor) {
  if (value instanceof constructor) {
    return true;
  }
  if (value) {
    const valueConstructor = value.constructor;
    if (valueConstructor && valueConstructor.name === constructor.name) {
      throw new Error(
        `Cannot use ${constructor.name} "${value}" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.`,
      );
    }
  }
  return false;
}
```

instanceOf.js
```
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = instanceOf;


// eslint-disable-next-line no-redeclare
function instanceOf(value, constructor) {
  if (value instanceof constructor) {
    return true;
  }
  if (value) {
    var valueConstructor = value.constructor;
    if (valueConstructor && valueConstructor.name === constructor.name) {
      throw new Error("Cannot use " + constructor.name + " \"" + value + "\" from another module or realm.\n\nEnsure that there is only one instance of \"graphql\" in the node_modules\ndirectory. If different versions of \"graphql\" are the dependencies of other\nrelied on modules, use \"resolutions\" to ensure only one version is installed.\n\nhttps://yarnpkg.com/en/docs/selective-version-resolutions\n\nDuplicate \"graphql\" modules cannot be used at the same time since different\nversions may have different capabilities and behavior. The data from one\nversion used in the function from another could produce confusing and\nspurious results.");
    }
  }
  return false;
} /**
   * Copyright (c) 2015-present, Facebook, Inc.
   *
   * This source code is licensed under the MIT license found in the
   * LICENSE file in the root directory of this source tree.
   *
   * 
   */

/**
 * A replacement for instanceof which includes an error warning when multi-realm
 * constructors are detected.
 */
```

**Comment by gaearon** [comment]
We definitely shouldn’t be copy pasting these. Can’t we just *move* types to a file called `index.js.flow` in the package root and then point imports to it?

**Comment by prometheansacrifice** [comment]
There are shared types involved. Which won't get linked unless they are
compiled some how.

**Comment by gaearon** [comment]
There might be ways to decouple them.

Also worth noting that types like Batch, FiberRoot, Fiber should be considered opaque and not leak to the renderers.

**Comment by prometheansacrifice** [comment]
> Also worth noting that types like Batch, FiberRoot, Fiber should be considered opaque and not leak to the renderers

I think I dont understand. FiberRoot (aliased as OpaqueRoot), which depends on Batch and Fiber, is returned by the `Reconciler`. They are needed to declare Reconciler's signature and are not exported. Have I gone wrong wrt this? Sorry if I'm stating the obvious or repeating myself.

> There might be ways to decouple them.
I'm still figuring. At the moment sharing these dependencies in a shared folder, possibly causing a lot of churn and moving around of types seems like the way to go. Should I go ahead? Or is there a better idea?

**Comment by jquense** [comment]
Dan is saying that generally these types should [opaque types](https://flow.org/en/docs/types/opaque-types/) but I believe the the tooling isn't quite there to use that feature (?). They are aliased in the types as "OpaqueHandle", etc so we don't actually export the Fiber type, and folks don't depend on the details of said type. 

As i understand the way graphql-js handles it it seems like the most straightforward way to handle it, if not a bit wonky. Regardless of how the types are exposed they'll need to be copied over to the root of the npm package so they are importable like `import type HostConfig from 'react-reconciler'`. Moving the types to common file would also work, but it's location may effect other pacakges so perhaps it's more complicated (maybe not)

**Comment by jquense** [comment]
Tho, I think in either case without decoupling all the types completely we will have problems because if flow has to require the types from another file, its not likely that file will be there, unless it's also completely types and copied over, and remains at the same relative path (so it's resolvable).

File: packages/react-reconciler/src/ReactFiberUpdateQueue.js [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 import type {Fiber} from './ReactFiber';
 import type {ExpirationTime} from './ReactFiberExpirationTime';
+import type {Update, UpdateQueue} from './ReactFiberReconcilerSharedTypes';
+export type {Update, UpdateQueue};
```
Comment (prometheansacrifice) [inline packages/react-reconciler/src/ReactFiberUpdateQueue.js]: I feel bad about this: is this okay?

File: scripts/flow/utils.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+// @noflow
```
Comment (prometheansacrifice) [inline scripts/flow/utils.js]: Heads up: A lot of flow fixmes.

**Comment by prometheansacrifice** [comment]
So I wrote a small babel script to extract the types. I hit a circular dependency in `ReactFiber`: ReactFiber -> ReactFiberUpdateQueue -> ReactFiber again. I moved a few types to `ReactFiberReconcilerSharedTypes.js` work around it. @gaearon Please give it a closer look. Flow coverage wasn't affected a lot (though I didn't do a thorough check).

If the babel script is fine, I'll update the build script to package it. Hope this helps :)

**Comment by prometheansacrifice** [comment]
@jquense What was it that you used to find all types erroneously marked `any`. Was it `flow coverage`?

File: scripts/rollup/packaging.js [hunk 3]
```diff
@@ -94,6 +94,13 @@ async function copyAllShims() {
   await Promise.all([copyWWWShims(), copyRNShims()]);
 }
 
+async function copyReactReconcilerTypes() {
+  await asyncCopyTo(
+    require.resolve('react-reconciler/src/ReactFiberReconcilerTypes.js'),
```
Comment (gaearon) [inline scripts/rollup/packaging.js]: Why can't we put these in `react-reconciler/index.js.flow` in the first place?

Comment (prometheansacrifice) [reply scripts/rollup/packaging.js]: Sure we can. This PR is not ready to be merged yet. But I wanted to know if I'm headed the right way with the babel script.

Comment (prometheansacrifice) [reply scripts/rollup/packaging.js]: Tried explaining in the comment earlier :)

**Comment by prometheansacrifice** [comment]
@gaearon Other than extracting the types directly to `index.js.flow` you mentioned earlier, does this PR need more work?

**Comment by gaearon** [comment]
I don’t know, can you start with that and then we’ll see?

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+  pendingContext: Object | null, // Determines if we should attempt to hydrate on the initial mount
+  +hydrate: boolean, // List of top-level batches. This list indicates whether a commit should be
+  // deferred. Also contains completion callbacks.
+  // TODO: Lift this into the renderer
+  firstBatch: Batch | null, // Linked-list of roots
+  nextScheduledRoot: FiberRoot | null,
+};
+export type Deadline = {timeRemaining: () => number};
+type OpaqueHandle = Fiber;
+type OpaqueRoot = FiberRoot;
```
Comment (jquense) [inline packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: these should be exported as well. They also should probably not be the actual `Fiber` type. The goal of this is that they are opaque to a user, meaning they track the right type, but the contents of the type aren't visible to a user of it, e.g react-dom-lite will pass around `OpaqueHandle` but should know or touch anything on it.

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    rootContainerInstance: C,
+    hostContext: CX,
+    internalInstanceHandle: OpaqueHandle,
+  ) => I,
+  appendInitialChild: (parentInstance: I, child: I | TI) => void,
+  finalizeInitialChildren: (
+    parentInstance: I,
+    type: T,
+    props: P,
+    rootContainerInstance: C,
```
Comment (jquense) [inline packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: this needs to be updated from the source again, the method sig changed here with my recent PR

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+  | ReactCall<any>
+  | ReactReturn<any>
+  | ReactPortal
+  | ReactText
+  | ReactFragment;
+type ReactFragment = ReactEmpty | Iterable<React$Node>;
+type ReactNodeList = ReactEmpty | React$Node;
+type ReactText = string | number;
+type ReactEmpty = null | void | boolean;
+type ReactCall<V> = {
```
Comment (jquense) [inline packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: I believe most of these types are already included in flow?

**Comment by prometheansacrifice** [comment]
To help you understand the PR:
I added a babel script that extracts all flow types. This runs and saves the results in `index.js.flow` just before npm packages are prepared. As I was writing the script, I noticed there was a circular dependency: I had to manually put them in ReactFiberReconcilerSharedTypes.js. 

Additionally ReactFiberReconcilerTypes.js is checked into the branch just so that the result of the babel script can be reviewed. I will remove them as it is not a part of the build process.

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js (hunk 5 above)
Comment (prometheansacrifice) [reply packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: Can you please point me to the change in your PR?

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js (hunk 6 above)
Comment (prometheansacrifice) [reply packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: Maybe. I don't know for sure. This file is auto generated. It's the result of the babel script. Right now it's present in the PR to only give a sense of what the output will look like.

File: packages/react-reconciler/src/ReactFiberReconcilerTypes.js (hunk 5 above)
Comment (jquense) [reply packages/react-reconciler/src/ReactFiberReconcilerTypes.js]: the PR is merged! check master for the updated signature

**Comment by prometheansacrifice** [comment]
I'll close this for now, but would love to attempt this again some time later. Hoping `gen-flow-files` files will help this time.

**Comment by gaearon** [comment]
We ended up getting rid of explicit types for that btw.

**Comment by prometheansacrifice** [comment]
> We ended up getting rid of explicit types for that btw.

I'm sorry I didn't understand that :)

**Comment by prometheansacrifice** [comment]
Did you mean you removed some explicit types so that gen-flow-files would work?

**Comment by gaearon** [comment]
I mean that we don't have `type FiberReconciler = ...` or `type HostConfig = ...` anymore.
Instead they're just module exports now.

**Pull request closed without merging**